qubes/storage/reflink.py

```diff
@@ -115,12 +118,37 @@ def included_in(self, app):
             [pool for pool in app.pools.values() if pool is not self],
             self.dir_path)
 
+
+def _unblock(method):
+    ''' Decorator transforming a synchronous volume method into a
+        coroutine that runs the original method in the event loop's
+        thread-based default executor, under a per-volume lock.
+    '''
+    @asyncio.coroutine
+    @functools.wraps(method)
+    def wrapper(self, *args, **kwargs):
+        with (yield from self._lock):  # pylint: disable=protected-access
+            return (yield from asyncio.get_event_loop().run_in_executor(
+                None, functools.partial(method, self, *args, **kwargs)))
+    return wrapper
```

Comment:
Generally, using run_in_executor should be considered last resort solution. If proper API exists, better use that.
Looking at the code, I think the only thing benefiting from this is `_copy_file` function. So, why not convert it to a proper coroutine (using [asyncio.create_subprocess_exec](https://docs.python.org/3/library/asyncio-subprocess.html)), then change this decorator to `_locked` or such (or inline lock handling)?

Reply:
On btrfs, even unlink and overwriting rename can take pretty long when operating on large fragmented snapshots, like easily 5-10 seconds under high load. So these would have to run in executors anyway. But then it just seems cleaner to do it once per method. There's probably not a lot to gain here with fine grained intra-method parallelization?

Another obstacle is that these helpers are also called from methods that are not allowed to be coros, so I guess there'd either have to be both a sync and an async version of each helper, or the sync methods would become responsible for scheduling async helper calls. :\

Reply:
Ew/thanks ;), I'll read up on that.

Reply:
> One concern is handling children processes from multiple threads (SIGCHLD etc) and asyncio (which I believe use signalfd) is potentially fragile.

I'm trying out a small commit (not pushed yet) that calls the FICLONE ioctl natively in `_copy_file()`, and only runs `cp --sparse=always` as a fallback. Too bad that Python doesn't seem to offer a native sparse copy function [yet](https://bugs.python.org/issue10016), if it did there wouldn't have to be any subprocesses in file-reflink at all.